### PR TITLE
Addition of Directional derivatives

### LINKED ISSUE #12416: Addition of Directional Derivative Function
Sympy, currently doesn't support the directional derivative function which is used for finding the components of a scalar field along a vector.

Directional Derivative functions are of importance these are used in vector mathematics as well as in differential geometry.

I hope that after this the issue has been solved we would be able to do the following:
```
>>> from sympy.vector import CoordSysCartesian, directional_derivative
>>> R = CoordSysCartesian('R')
>>> f1 = R.x*R.y*R.z
>>> v1 = 3*R.i + 4*R.j + R.k
>>> directional_derivative(s1, v1, R)
3*R.y*R.z + 4*R.x*R.z + R.x*R.y
```

### PR DESCRIPTION
**Pull request by mohit3011**
Fixes #12416

### CODE REVIEW

**Comment by mohit3011** [comment]
@jksuom , @asmeurer   please review this PR . Thanks

File: sympy/vector/functions.py [hunk 1]
```diff
@@ -214,6 +214,38 @@ def gradient(scalar, coord_sys):
 
     return coord_sys.delop(scalar).doit()
 
+def directional_derivative(scalar, vect, coord_sys):
+    """
+    Returns the directional derivative of a scalar field computed along a  given vector
```
Comment (szymag) [inline sympy/vector/functions.py]: Here you have double white space, after "along a".
If you want to correct this, you shouldn't push new commit. You can use `git commit --amend`. It overrides before one.

Comment (mohit3011) [inline sympy/vector/functions.py]: @szymag , thanks for reviewing it. I will correct the mistake.

**Comment by szymag** [comment]
I have some doubts about your solution.
 
I think that vector, which you take as parameter, should be normalized. Here, you can find some information about it (here or in the next film):
https://www.khanacademy.org/math/multivariable-calculus/multivariable-derivatives/gradient-and-directional-derivatives/v/directional-derivative

Directional derivative is a function, but you return vector. More explicitly `<class 'sympy.core.add.Add'>`. It may be complicated for user, who also want to calculate value of functional derivative in some point. 

What do you think?

**Comment by mohit3011** [comment]
Also quoting from Wikipedia's Definition 

> In the context of a function on a Euclidean space, some texts restrict the vector v to being a unit vector. Without the restriction, this definition is valid in a broad range of contexts.

Therefore I have not kept the condition of `V` being a unit vector so that it could be used in broader sense.

**Comment by szymag** [comment]
You always should allowed user to put any vector, but then normalized them. 
My argumentation here is that, when you put two different vector with the same direction but different length, you get different result. 

I'm not going to persist on my idea, it is just convention. It will be better, if someone from maintainers takes a look.

**Comment by mohit3011** [comment]
ping @jksuom  @asmeurer  , please review this PR

**Comment by siefkenj** [comment]
The directional derivative is unfortunately named, but you should allow **any** vector (including the zero vector) as input.  That way, the directional derivative is a linear function in its direction for differentiable functions.  Some textbooks work around the issue by defining the directional derivative only for unit vectors and leaving it the term "directional derivative" unspecified for non-unit vectors.

However, your code will only produce the directional derivative for a differentiable function.  The true definition of the directional derivative of `f` at the point `x` in the direction `v` is `lim_{h -> 0}( f(x+hv)-f(x))/h` which is only equal to `grad f(x) . v` if `f` is differentiable.  Does sympy have a differentiability test?

**Comment by mohit3011** [comment]
@siefkenj 

> The directional derivative is unfortunately named, but you should allow any vector (including the zero vector) as input. 

As you said this functions allows any vector as input.

>  directional derivative of f at the point x in the direction v is lim_{h -> 0}( f(x+hv)-f(x))/h which is only equal to grad f(x) . v if f is differentiable.

I have used the existing Gradient function in Sympy for DIrectional Derivative so it must be checking the conditions of differentiability as Gradient function further calls the existing Derivative function of Sympy.

File: sympy/vector/functions.py [hunk 2]
```diff
@@ -214,6 +214,38 @@ def gradient(scalar, coord_sys):
...
+    >>> f1 = R.x*R.y*R.z
+    >>> v1 = 3*R.i + 4*R.j + R.k
+    >>> directional_derivative(f1, v1, R)
+    R.x*R.y + 4*R.x*R.z + 3*R.y*R.z
+    >>> f2 = 5*R.x**2*R.z
+    >>> directional_derivative(f2, v1, R)
+    5*R.x**2 + 30*R.x*R.z
+
+    """
+    return gradient(scalar,coord_sys).dot(vect).doit()
```
Comment (Upabjojr) [inline sympy/vector/functions.py]: This is unrelated to this PR, but I believe `gradient` should be called as `gradient(scala)` without the `coord_sys` variable.

Comment (szymag) [inline sympy/vector/functions.py]: Not really, look at documentation:
http://docs.sympy.org/dev/modules/vector/fields.html
We need manually put information about coordinate system. For me it is also strange

Comment (Upabjojr) [inline sympy/vector/functions.py]: I know that, that's why we need to modify the gradient function.

Comment (szymag) [inline sympy/vector/functions.py]: Sorry, I misunderstand your comment a bit.
> I know that, that's why we need to modify the gradient function.
 
I agree with you, but now it isn't so easy. If I correctly understand code, `vector` has no information about his coordinate system, we need forward this information explicitly to function (like gradient). 
I am writing proposal, where I also include this information, and I would like to fix this issue.

Comment (mohit3011) [inline sympy/vector/functions.py]: @Upabjojr , It won't be that easy now as this would mean changing the whole sytem of deloperator.py and functions.py.

Comment (mohit3011) [inline sympy/vector/functions.py]: It is because of the function `gradient` in functions.py depends on `delop` which is in deloperator.py

Comment (mohit3011) [inline sympy/vector/functions.py]: If it is not a problem then can we merge it for now. The reason being in `deloperator.py` we need the coordinate system as the class objects are defined by the reference of the coordinate system variables. Thus we need to pass the coordinate system for reference purposes.

Comment (Upabjojr) [inline sympy/vector/functions.py]: Yes, I wrote that this issue is not related to this PR.

Comment (Upabjojr) [inline sympy/vector/functions.py]: > If I correctly understand code, vector has no information about his coordinate system,

It has:
```python
In [6]: c = CoordSysCartesian('rect')

In [7]: c
Out[7]: rect

In [8]: c.x
Out[8]: rect_x

In [9]: c.x.args
Out[9]: (rect.x, 0, rect, rectₓ, \mathbf{{x}_{rect}})

In [10]: c.x.args[2]
Out[10]: rect

In [11]: c.x.args[2] == c
Out[11]: True
```

Comment (mohit3011) [inline sympy/vector/functions.py]: @Upabjojr  , if there is no problem then could this PR be merged.

Comment (mohit3011) [inline sympy/vector/functions.py]: @Upabjojr , in reply to your last comment. I would like to point out that in whenever we call the `gradient` function from vector/functions.py then it calls the class `CoordSysCartesian` in vector/coordsysrect.py int his class `delop` function is called. which further calls `Del` function in deloperator.py . Now, the class `CoordSysCartesian`  requires the name of Coordinate system for assigning purposes so that's we are passing coordinate system in gradient function.

Comment (Upabjojr) [inline sympy/vector/functions.py]: Yeah, it needs some cleanup.

Comment (mohit3011) [inline sympy/vector/functions.py]: @Upabjojr , please merge this PR. Thanks

File: sympy/vector/functions.py [hunk 3]
```diff
@@ -214,6 +214,38 @@ def gradient(scalar, coord_sys):
...
+        The coordinate system to calculate the gradient in
+
+    Examples
+    ========
+
+    >>> from sympy.vector import CoordSysCartesian, directional_derivative
+    >>> R = CoordSysCartesian('R')
+    >>> f1 = R.x*R.y*R.z
+    >>> v1 = 3*R.i + 4*R.j + R.k
+    >>> directional_derivative(f1, v1, R)
```
Comment (Upabjojr) [inline sympy/vector/functions.py]: It's the same problem as in the gradient function. I don't like the `R` parameter in the API.

Comment (mohit3011) [inline sympy/vector/functions.py]: @Upabjojr ,  I understand what you mean, but as I told you earlier it would be a very large amount of work to clean up the whole coordinate system with each function. So for the existing system this is the best possible way

Comment (mohit3011) [inline sympy/vector/functions.py]: Maybe the work for redifining the whole system can done in other PRs

Comment (Upabjojr) [inline sympy/vector/functions.py]: No, it's simple. Use `.atoms( )`.

Comment (mohit3011) [inline sympy/vector/functions.py]: What does .atoms() does ?

Comment (mohit3011) [inline sympy/vector/functions.py]: @Upabjojr , Secondly, I have taken the coordinate system 'R' just because I'm using the gradient function in directional derivative.

Comment (Upabjojr) [inline sympy/vector/functions.py]: Maybe you can also extract the coordinate system object from vector.

Comment (mohit3011) [inline sympy/vector/functions.py]: @Upabjojr , maybe we can extract that information but then it would create a problem that it would be out of convention. What I mean is that in the existing system the user has to to give cordinate system as input in every function , now if I create a function where user doesn't have to give the coordinate system as input then it would create confusion and disturb the conventions.

Comment (mohit3011) [inline sympy/vector/functions.py]: Every time user wishes to use this function then he/she has to use a syntax which would be different from other functions.

Comment (Upabjojr) [inline sympy/vector/functions.py]: Let's start writing the new functions with an intuitive syntax. This can be fixed by adding one line to extract the coordinate system and then calling the gradient.

Comment (mohit3011) [inline sympy/vector/functions.py]: Okay , let's do it

**Comment by Upabjojr** [comment]
vector.atoms(CoordSysCartesian)

**Comment by mohit3011** [comment]
@Upabjojr , it has passed all Tests except one which has error due to exceeded time limit.

**Comment by mohit3011** [comment]
@Upabjojr  , please review this PR thanks

**Comment by mohit3011** [comment]
@Upabjojr , Thanks
